### PR TITLE
Add guides for Tizen Web ML API

### PR DESCRIPTION
**Pull request by pwasowski2**
### Change Description

Add guides for Tizen Web ML API

### API Changes

- ACR: TWDAPI-273
- ACR: TWDAPI-274

This is continuation of work from this PR https://github.com/Samsung/tizen-docs/pull/1432

### CODE REVIEW

**Comment by pwasowski2** [comment]
@nawab-samsung, I kindly ask you for review - I have fixed all the problems you have found in review of https://github.com/Samsung/tizen-docs/pull/1432

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+    ```javascript
+    videotestsrcNode.setProperty('pattern', 'ENUM', 18);
+    ```
+
+ > [!NOTE]
+ > You can also set the `pattern` value in pipeline description, for example, `videotestsrc pattern=18 ! tizenwlsink`
+
+## Input data from application
+
+The pipeline input can be generated by specialized nodes, for example `videotestsrc` or by the application code, feeding `appsrc` nodes.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/pipeline.md:221]:
```suggestion 
The pipeline input can be generated by specialized nodes, for example `videotestsrc` or by the application code feeding the `appsrc` nodes.

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+
+4. Set the desired valve state:
+
+    ```javascript
+    valve.setOpen(false); // stop the data flow
+    ```
+
+## Write custom data filters
+
+The data flowing through the pipeline can be transformed in application, using  `CustomFilter` callbacks.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/pipeline.md:375]:
```suggestion
The data flowing through the pipeline can be transformed in application using `CustomFilter` callbacks.

File: docs/application/web/guides/machine-learning/tensors.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+2. `TensorsInfo` object can store information such as name, data type, and dimensions:
+
+    ```javascript
+    tensorsInfo.addTensorInfo("tensor", "UINT8", [2, 2]);
+    ```
+
+ > [!NOTE]
+ >
+ > One `TensorsInfo` object can store information on up to 16 tensors.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:46]:
```suggestion
 > [!NOTE]
 > A single `TensorsInfo` object can store information of up to 16 tensors.

File: docs/application/web/guides/machine-learning/tensors.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    ```javascript
+    // fetch one element
+    var rawDataOne = tensorsData.getTensorRawData(0, [0, 0], [1, 1]);
+    // fetch first row
+    var rawDataRow = tensorsData.getTensorRawData(0, [0, 0], [-1, 1]);
+    ```
+
+ > [!NOTE]
+ > To gather more information about specifying location and size of the raw data, see `TensorsData.getTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::getTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::getTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::getTensorRawData) applications) to gather more information about specifying location and size of raw data.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:104]:
```suggestion
 > [!NOTE]
 > To gather more information about specifying location and size of the raw data, see `TensorsData.getTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::getTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::getTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::getTensorRawData) applications) .

File: docs/application/web/guides/machine-learning/tensors.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    ```javascript
+    // set only one element
+    tensorsData.setTensorRawData(0, [7], [0, 0], [1, 1]);
+    // set first row
+    tensorsData.setTensorRawData(0, [4, 5], [0, 0], [-1, 1]);
+    ```
+
+ >  [!NOTE]
+ > To gather more information about specifying location and size of the raw data, see `TensorsData.setTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::setTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::setTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::setTensorRawData) applications) to gather more information about specifying location and size of raw data.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:122]:
```suggestion
>  [!NOTE]
 > To gather more information about specifying location and size of the raw data, see `TensorsData.setTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::setTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::setTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::setTensorRawData) applications).

**Review by nawab-samsung — CHANGES_REQUESTED** [review]
@pwasowski2 
Kindly incorporate the comments. 
A few comments have not been applied from previous review.
Thanks.

File: docs/application/web/guides/machine-learning/overview.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# Machine Learning
+
+Machine Learning features help you to handle Neural Network Frameworks (NNFWs), like [TensorFlow and TensorFlow Lite](https://tensorflow.org).
+
+You can use the following machine learning features in your Web applications:
+
+- [Tensors Management](./tensors.md)
+
+  You can create and manage `TensorsInfo` and `TensorsData` objects. They allow you to exchange data with neural models.
```
[the remark this answers is not in the document]
Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/overview.md:9]: Done

File: docs/application/web/guides/machine-learning/overview.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# Machine Learning
```
[the remark this answers is not in the document]
Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/overview.md:1]: Sure, I'll change headers in this file to sentence case

**Comment by pwasowski2** [comment]
@nawab-samsung, thanks for the review :) I have fixed the issues you found.

I have commented one of your suggestions - could you please take a look at my comment?

File: docs/application/web/guides/machine-learning/tensors.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+The main features of the Machine Learning API include:
+
+- Checking support of a specific neural network framework
+
+  You can [check NNFW availability](#check-nnfw-availability) using `checkNNFWAvailability`.
+
+- Creating tensor's structure.
+
+  You can use `TensorsInfo` object to [manage tensor information](#manage-tensor-information).
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:13]:
```suggestion
  You can use the `TensorsInfo` object to [manage the tensor information](#manage-tensor-information).

File: docs/application/web/guides/machine-learning/tensors.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+- Reading and writing raw data.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:15]: To maintain consistency please remove the period
```suggestion
- Reading and writing raw data

File: docs/application/web/guides/machine-learning/tensors.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+  You can [read and write tensor data](#manage-tensor-data) using `TensorsData` object.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:17]:
```suggestion
  You can [read and write tensor data](#manage-tensor-data) using the `TensorsData` object.

File: docs/application/web/guides/machine-learning/tensors.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+## Check NNFW availability
+
+To check whether specific Neural Network Framework (NNFW) is supported you can use `checkNNFWAvailability`:
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:21]:
```suggestion
To check whether specific Neural Network Framework (NNFW) is supported, you can use `checkNNFWAvailability`:

**Review by nawab-samsung — CHANGES_REQUESTED** [review]
@pwasowski2 A few missed comments, please incorporate. est seems to be fine,

File: docs/application/web/guides/machine-learning/single.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+      model.close();
+    }
+    tizen.ml.single.openModelAsync("documents/mobilenet_v1_1.0_224_quant.tflite", successCallback, errorCallback);
+    ```
+
+## Invoke
+
+To invoke the neural network model, you need to create the `TensorsData` instance to pass the input data to the model. You can find the details about using `TensorsData` interface in [Tensors Management](./tensors.md) guide.
+
+1. Create `TensorsData` object based on information from model:
```
[the remark this answers is not in the document]
Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/single.md:102]: Ok, I'll just remove an unnecessary space.

```suggestion
1. Create the `TensorsData` object based on the information from the model:
```

**Review by pwasowski2 — COMMENTED** [review]
@nawab-samsung, thank you for the review :) I'll commit your suggestions

**Comment by nawab-samsung** [comment]
@pwasowski2 Thanks for incorporating the comments.

**Review by nawab-samsung — CHANGES_REQUESTED** [review]
@pwasowski2 
Thanks!
A few comments for ensuring consistency & following style guide rules. Kindly incorporate the comments, following which I will approve the PR as no more changes are needed. But please ensure you don't miss any comment in the current commit as was the case previously.

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+                              '! tensor_converter ' +
+                              '! tensor_filter framework=custom-easy model=flattenFilter ' +
+                              '! fakesink';
+    var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
+    ```
+When you start the pipeline, the `flattenFilter` transforms 3-dimensional tensors into 1-dimensional vector.
+
+## Use saved models
+
+Using machine learning models trained with popular frameworks like [TensorFlow](https://tensorow.org) is one of the main use cases for NNStreamer pipelines.
```
[the remark this answers is not in the document]
Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/pipeline.md:422]: Done

**Comment by pwasowski2** [comment]
@nawab-samsung, thanks for the review, I have applied your suggestions.

File: docs/application/web/guides/machine-learning/single.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+      // Do inference here
+
+      model.close();
+    }
+    tizen.ml.single.openModelAsync("documents/mobilenet_v1_1.0_224_quant.tflite", successCallback, errorCallback);
+    ```
+
+## Invoke
+
+To invoke the neural network model, you need to create the `TensorsData` instance to pass the input data to the model. You can find the details about using `TensorsData` interface in [Tensors Management](./tensors.md) guide.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/single.md:99]:
```suggestion
To invoke the neural network model, you need to create the `TensorsData` instance to pass the input data to the model. You can find the details about using `TensorsData` interface in [Tensors Management](./tensors.md) guide:

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+    console.log(videotestsrcNode.name + '\'s pattern: ' + pattern); // 'srcx's pattern: 0';
+    ```
+
+4. The current `pattern` is `0`, which translates to `frames` pattern, according to `gst-inspect-1.0` output. Change it to `ball` by setting property value to `18`:
+     
+    ```javascript
+    videotestsrcNode.setProperty('pattern', 'ENUM', 18);
+    ```
+
+ > [!NOTE]
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/pipeline.md:216]: This note is for point 4 so we need to indent it right: 
```suggestion
    > [!NOTE]
    > You can also set the `pattern` value in pipeline description, for example, `videotestsrc pattern=18 ! tizenwlsink`

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+
+      return 0;
+    };
+    tizen.ml.pipeline.registerCustomFilter('flattenFilter', flattenFilter, inputInfo, outputInfo);
+    ```
+
+ > [!NOTE]
+ > `inputData` and `outputData` passed to `flattenFilter` are different than other `TensorsData` objects.
+ > These objects cannot be disposed and are only valid within the callback. You have to copy the objects manually such that you can use these objects outside the callback.
+ > `inputData` is read-only and `outputData` is initialized with random values.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/pipeline.md:406]:
```suggestion
     > [!NOTE]
     > `inputData` and `outputData` passed to `flattenFilter` are different than other `TensorsData` objects.
     > These objects cannot be disposed and are only valid within the callback. You have to copy the objects     
         manually such that you can use these objects outside the callback.
     > `inputData` is read-only and `outputData` is initialized with random values.

File: docs/application/web/guides/machine-learning/tensors.md [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+Machine Learning Web API in Tizen requires `TensorsInfo` and `TensorsData` objects. These objects are used to perform calculation with neural network models.
+
+The main features of the Machine Learning API include:
+
+- Checking support of a specific neural network framework
+
+  You can [check NNFW availability](#check-nnfw-availability) using `checkNNFWAvailability`.
+
+- Creating tensor's structure.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:11]: To maintain consistency no period. 
```suggestion
- Creating tensors structure

File: docs/application/web/guides/machine-learning/tensors.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+    ```
+
+2. `TensorsInfo` object can store information such as name, data type, and dimensions:
+
+    ```javascript
+    tensorsInfo.addTensorInfo("tensor", "UINT8", [2, 2]);
+    ```
+
+ > [!NOTE]
+ > A single `TensorsInfo` object can store information about up to 16 tensors.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:45]:
```suggestion
    > [!NOTE]
    > A single `TensorsInfo` object can store information about up to 16 tensors.

File: docs/application/web/guides/machine-learning/tensors.md [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+
+4. To clone and compare `TensorsInfo` object:
+
+    ```javascript
+    var clone = tensorsInfo.clone();
+    var isEqual = tensorsInfo.equals(clone);
+    ```
+
+ > [!NOTE]
+ > Use `equals()` to compare `TensorInfo` objects. Comparisons must not be done with built-in `==`, `===`, `!=`, `!==` operators which may return false results for this data type.
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:63]:
```suggestion
     > [!NOTE]
     > Use `equals()` to compare `TensorInfo` objects. Comparisons must not be done with built-in `==`, `===`, !=`, `!==` operators which may return false results for this data type.

File: docs/application/web/guides/machine-learning/tensors.md [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+    ```javascript
+    // fetch one element
+    var rawDataOne = tensorsData.getTensorRawData(0, [0, 0], [1, 1]);
+    // fetch first row
+    var rawDataRow = tensorsData.getTensorRawData(0, [0, 0], [-1, 1]);
+    ```
+
+ > [!NOTE]
+ > To gather more information about specifying location and size of the raw data, see `TensorsData.getTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::getTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::getTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::getTensorRawData) applications).
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:103]:
```suggestion
    > [!NOTE]
    > To gather more information about specifying location and size of the raw data, see `TensorsData.getTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::getTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::getTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::getTensorRawData) applications).

File: docs/application/web/guides/machine-learning/tensors.md [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+
+    ```javascript
+    // set only one element
+    tensorsData.setTensorRawData(0, [7], [0, 0], [1, 1]);
+    // set first row
+    tensorsData.setTensorRawData(0, [4, 5], [0, 0], [-1, 1]);
+    ```
+
+ >  [!NOTE]
+ > To gather more information about specifying location and size of the raw data, see `TensorsData.setTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::setTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::setTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::setTensorRawData) applications).
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:121]:
```suggestion
     >  [!NOTE]
     > To gather more information about specifying location and size of the raw data, see `TensorsData.setTensorRawData` (in [mobile](../../api/latest/device_api/mobile/tizen/ml.html#TensorsData::setTensorRawData), [wearable](../../api/latest/device_api/wearable/tizen/ml.html#TensorsData::setTensorRawData), and [TV](../../api/latest/device_api/tv/tizen/ml.html#TensorsData::setTensorRawData) applications).

**Review by nawab-samsung — CHANGES_REQUESTED** [review]
The notes need to be aligned appropriately, sorry, I had missed this in the previous comments.

File: docs/application/web/guides/machine-learning/tensors.md (hunk 19 above)
Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/tensors.md:63]: Ok, I'll just add a missing "`":

```suggestion
     > [!NOTE]
     > Use `equals()` to compare `TensorInfo` objects. Comparisons must not be done with built-in `==`, `===`, `!=`, `!==` operators which may return false results for this data type.
```

**Comment by pwasowski2** [comment]
Thanks for the review, @nawab-samsung :) I have applied your suggestions

File: docs/application/web/guides/machine-learning/tensors.md [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+
+The `TensorsData` object keeps data value of the tensors.
+
+1. To get specific data object from `TensorsData`, use `getTensorRawData`:
+
+    ```javascript
+    var rawData = tensorsData.getTensorRawData(0);
+    ```
+
+- You can also specify the location and size of data you want to get:
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:93]: Just a doubt: 

line 93 & following code:

Is it a sub-step of 1?

If yes, we need to move the 93-100, & 112- 118 to four spaces right.

**Review by nawab-samsung — COMMENTED** [review]
@pwasowski2 All seems well, Just one comment. I need your suggestion.

File: docs/application/web/guides/machine-learning/tensors.md (hunk 22 above)
Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/tensors.md]: Yes, these are substeps of 1 and 2. I've indented the lines.

**Comment by pwasowski2** [comment]
@nawab-samsung, thank you for the review :) I have fixed the indentation

**Comment by nawab-samsung** [comment]
@pwasowski2 Thanks for incorporating comments & clearing my doubts.

**Review by nawab-samsung — APPROVED** [review]
@pwasowski2 Thanks for your cooperation. 
https://docs1.stg.tizen.org/staging/1437/application/web/guides/machine-learning/overview/

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 23]
```diff
@@ -0,0 +1,479 @@
...
+
+  You can [use CustomFilter API to write custom data processing routines in JS](#write-custom-data-filters).
+
+- Use saved models
+
+  You can use [`tensor_filter` element to read models trained with popular machine learning frameworks](#use-saved-models).
+
+## Prerequisites
+
+To access files, camera or recorder using the Pipeline API (in [mobile](../../api/latest/device_api/mobile/tizen/ml_pipeline.html), [wearable](../../api/latest/device_api/wearable/tizen/ml_pipeline.html), and [tv](../../api/latest/device_api/tv/tizen/ml_pipeline.html) applications), the application has to define proper privileges in its `config.xml`:
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/pipeline.md:51]: You missed this comment, tv-->TV for the sake of consistency across guides.
```suggestion
To access files, camera or recorder using the Pipeline API (in [mobile](../../api/latest/device_api/mobile/tizen/ml_pipeline.html), [wearable](../../api/latest/device_api/wearable/tizen/ml_pipeline.html), and [TV](../../api/latest/device_api/tv/tizen/ml_pipeline.html) applications), the application has to define proper privileges in its `config.xml`:

File: docs/application/web/guides/machine-learning/pipeline.md [hunk 24]
```diff
@@ -0,0 +1,479 @@
...
+    ```javascript
+    pipeline.dispose();
+    ```
+
+The disposed pipelines enter the `NULL` state, and cannot be restarted. Any attempt of calling any method of a disposed pipeline results in `NotFoundError` exception.
+
+## Observe pipeline state and respond to its changes
+
+Pipeline state reflects what the pipeline is doing right now and defines which methods can be called.
+To see how pipelines transition between different states, see the API reference (for [mobile](../../api/latest/device_api/mobile/tizen/ml_pipeline.html#PipelineState), [wearable](../../api/latest/device_api/wearable/tizen/ml_pipeline.html#PipelineState), and [tv](../../api/latest/device_api/tv/tizen/ml_pipeline.html#PipelineState) applications).
```
Comment (nawab-samsung) [inline docs/application/web/guides/machine-learning/pipeline.md:111]:
```suggestion
To see how pipelines transition between different states, see the API reference (for [mobile](../../api/latest/device_api/mobile/tizen/ml_pipeline.html#PipelineState), [wearable](../../api/latest/device_api/wearable/tizen/ml_pipeline.html#PipelineState), and [TV](../../api/latest/device_api/tv/tizen/ml_pipeline.html#PipelineState) applications).

**Review by nawab-samsung — CHANGES_REQUESTED** [review]
@pwasowski2 Please commit these changes, shall approve afterwards. 
This is for maintaining consistency.

**Comment by pwasowski2** [comment]
@nawab-samsung, I have commited the suggestions

**Comment by nawab-samsung** [comment]
@pwasowski2 Thanks a lot!

**Review by nawab-samsung — APPROVED** [review]
@pwasowski2 Thanks for your support & cooperation. 
Approving the PR.

**Comment by annie-samsung** [comment]
@mijin85cho, same question, since 6.5 branch is merged to master, should the branch for this PR be changed to master?

File: docs/application/web/guides/machine-learning/tensors.md [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+
+3. Ensure to dispose off the `TensorsData` objects when you do not need them anymore:
+
+    ```javascript
+    tensorsData.dispose();
+    ```
+
+## Related information
+
+- Dependencies
```
Comment (annie-samsung) [inline docs/application/web/guides/machine-learning/tensors.md:131]: @pwasowski2, @nawab-samsung I missed this bit, could you check why are the dependencies repeated twice here

Comment (pwasowski2) [reply docs/application/web/guides/machine-learning/tensors.md:131]: I see, @nawab-samsung has already prepared a fix for this: https://github.com/Samsung/tizen-docs/commit/bcbdae1bf914dab1bc09d529e74801d1d4e2884c

Thanks!

**Pull request merged**